v2/api/network/customizations/load_balancer_extension.go

```diff
@@ -120,33 +113,137 @@ func fuzzySetInboundNatRules(lb genruntime.ARMResourceSpec, inboundNatRules []ge
+			err = errors.Errorf("caught panic: %s", x)
+		}
+	}()
+
+	inboundNatRulesField, err := getInboundNatRulesField(lb)
+	if err != nil {
+		return err
 	}
 
 	elemType := inboundNatRulesField.Type().Elem()
-	inboundNatRuleSlice := reflect.MakeSlice(inboundNatRulesField.Type(), 0, 0)
+	inboundNatRulesSlice := reflect.MakeSlice(inboundNatRulesField.Type(), 0, 0)
 
-	for _, inboundNatRule := range inboundNatRules {
-		embeddedInboundNatRule := reflect.New(elemType)
-		err := fuzzySetInboundNatRule(inboundNatRule, embeddedInboundNatRule)
+	for _, inboundNatRule := range azureInboundNatRules {
+		embeddedInboundNatRules := reflect.New(elemType)
+		err = fuzzySetInboundNatRule(inboundNatRule, embeddedInboundNatRules)
 		if err != nil {
 			return err
 		}
 
-		inboundNatRuleSlice = reflect.Append(inboundNatRuleSlice, reflect.Indirect(embeddedInboundNatRule))
+		inboundNatRulesSlice = reflect.Append(inboundNatRulesSlice, reflect.Indirect(embeddedInboundNatRules))
 	}
 
-	// Now do the assignment. We do it differently here, as we need to make sure to retain current/updated/deleted InboundNatRules present on the LoadBalancer.
-	inboundNatRulesField.Set(reflect.AppendSlice(inboundNatRulesField, inboundNatRuleSlice))
+	inboundNatRulesSlice, err = mergeNatRules(inboundNatRulesField, inboundNatRulesSlice)
+	if err != nil {
+		return errors.Wrapf(err, "failed to merge NAT rules")
+	}
+
+	inboundNatRulesField.Set(inboundNatRulesSlice)
 
 	return nil
 }
 
-func fuzzySetInboundNatRule(inboundNatRule genruntime.ARMResourceSpec, embeddedInboundNatRule reflect.Value) error {
+func mergeNatRules(inboundNatRulesField reflect.Value, azureInboundNatRules reflect.Value) (reflect.Value, error) {
+	if inboundNatRulesField.Len() == 0 {
+		return azureInboundNatRules, nil
+	}
+
+	resultSlice := reflect.MakeSlice(inboundNatRulesField.Type(), 0, 0)
+
+	// The result slice should take every item from inboundNatRulesField, and merge in anything from azureInboundNatRules
+	// with a different name
+	for i := 0; i < inboundNatRulesField.Len(); i++ {
+		inboundNatRule := inboundNatRulesField.Index(i)
+		resultSlice = reflect.Append(resultSlice, inboundNatRule)
+	}
+
+	for i := 0; i < azureInboundNatRules.Len(); i++ {
+		inboundNatRule := azureInboundNatRules.Index(i)
+		newRuleName, err := getNameField(inboundNatRule)
+		if err != nil {
+			return reflect.Value{}, errors.Wrapf(err, "failed to get name for new rule")
+		}
+		foundExistingRule := false
+
+		for j := 0; j < inboundNatRulesField.Len(); j++ {
+			existingInboundNatRule := inboundNatRulesField.Index(j)
+			var existingName reflect.Value
+			existingName, err = getNameField(existingInboundNatRule)
+			if err != nil {
+				return reflect.Value{}, errors.Wrapf(err, "failed to get name for existing rule")
+			}
+
+			if existingName.String() == newRuleName.String() {
+				foundExistingRule = true
+				break
+			}
+		}
+
+		if !foundExistingRule {
+			resultSlice = reflect.Append(resultSlice, inboundNatRule)
+		}
```

Comment:
Is order significant? This puts all the extras at the end of the slice, so any that are known by ASO will be promoted in front of the others.

Reply:
Order doesn't seem to matter. The networking RP seems to return the collection in order of creation time, so the oldest NatRules are first and the newer ones are last. Whatever order you supply rules to it in, that doesn't change.

For our purposes because we're doing a merge here based on name, it doesn't make a different to us I don't think, so the order we end up with here should be OK.